### ISSUE TITLE
[no-unnecessary-type-assertion] false positive when asserting via non-null assertion operator

### ISSUE THREAD

**Issue by vkrol** [typescript-eslint/typescript-eslint#529]
**Repro**
https://github.com/vkrol/typescript-eslint-no-unnecessary-type-assertion-undefined

```js
module.exports = {
	parser: "@typescript-eslint/parser",
	parserOptions: {
		project: "tsconfig.json",
	},
	plugins: ["@typescript-eslint"],
	rules: {
		"@typescript-eslint/no-unnecessary-type-assertion": "error",
	}
};
```

```TS
declare function foo(str?: string): void;
declare const str: string | null;

foo(str!);
```
**Expected Result**
No errors.
**Actual Result**
Error:
```
D:\Projects\typescript-eslint-no-unnecessary-type-assertion-undefined\index.ts
  4:5  error  This assertion is unnecessary since the receiver accepts the original type of the expression  @typescript-eslint/no-unnecessary-type-assertion

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```
**Additional Info**
The change that introduced this regression: https://github.com/typescript-eslint/typescript-eslint/pull/478.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.9.0` |
| `@typescript-eslint/parser`        | `1.9.0` |
| `TypeScript`                       | `3.4.5` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.4.1` |

**Comment by vkrol** [comment]
@bradzacher I think that this change https://github.com/typescript-eslint/typescript-eslint/pull/478 introduced this regression.

**Comment by bradzacher** [comment]
no doubt - that PR added a lot of code and additional checks and had to work around uninitialised variables (#453).

this bug only occurs when `strictNullChecks` is turned on.
